### ISSUE TITLE
Endless recursion in "networkx" decorators leading to segfault

### ISSUE THREAD

**Issue by AntonOvsyannikov** [Nuitka/Nuitka#2157]
* Nuitka version, full Python version, flavor, OS, etc
Minimal example built with docker, but it also can be reproduced in different environments on mac, windows with same issue.
Also reproduced on **Commercial** 2.0.0 (please see below)
```
# python -m nuitka --version
1.5.5
Commercial: None
Python: 3.10.11 (main, Apr  5 2023, 23:35:11) [GCC 10.2.1 20210110]
Flavor: Unknown
Executable: /usr/local/bin/python
OS: Linux
Arch: x86_64
Distribution: Debian 11
Version C compiler: /usr/bin/gcc (gcc).
```
* How did you install Nuitka and Python
python from official docker image, no venv
`pip install nuitka==1.5.5` 

* The specific PyPI names and versions

```
# python -m pip freeze
networkx==2.8.8
Nuitka==1.5.5
ordered-set==4.1.0
zstandard==0.20.0
```
* Short, Self Contained, Correct, Example
Place `Dockerfile` in some directory
```
FROM python:3.10
WORKDIR /srv
ENV PYTHONUNBUFFERED=1

RUN apt update
RUN apt install patchelf

RUN printf 'nuitka==1.5.5\nnetworkx==2.8.8' >> requirements.txt


RUN python -m pip install -r requirements.txt

RUN printf '\
import networkx as nx\n\
g = nx.MultiDiGraph()\n\
g.add_edge(0, 1)\n\
g.add_edge(1, 2)\n\
print("Edges in original graph:", len(g.edges))\n\
gg = nx.MultiDiGraph(nx.subgraph_view(g, filter_edge=lambda u, v, k: u > 0))\n\
print("Edges in filtered graph:", len(gf.edges))\n\
' >> main.py

RUN python -m nuitka --standalone main.py

CMD "./main.dist/main.bin"
```
than type
`docker build -t nuitka_nx .`
```
...
Step 9/10 : RUN python -m nuitka --standalone main.py
 ---> Running in 028bc52e2727
Nuitka-Options:INFO: Used command line options: --standalone main.py
Nuitka:INFO: Starting Python compilation with Nuitka '1.5.5' on Python '3.10' commercial grade 'not installed'.
Nuitka:INFO: Completed Python level compilation and optimization.
Nuitka:INFO: Generating source code for C backend compiler.
Nuitka:INFO: Running data composer tool for optimal constant value handling.
Nuitka:INFO: Running C compilation via Scons.
Nuitka-Scons:INFO: Backend C compiler: gcc (gcc).
Nuitka-Scons:INFO: Backend linking program with 282 files (no progress information available).
Nuitka-Scons:WARNING: You are not using ccache.
Nuitka:INFO: Keeping build directory 'main.build'.
Nuitka:INFO: Successfully created 'main.dist/main.bin'.
Removing intermediate container 028bc52e2727
...
```
then `docker run nuitka_nx`
```
> docker run nuitka_nx
Edges in original graph: 2
Segmentation fault
```
Running this code in python produces correct output
`main.py`
```python
import networkx as nx

g = nx.MultiDiGraph()
g.add_edge(0, 1)
g.add_edge(1, 2)
print("Edges in original graph:", len(g.edges))

gg = nx.MultiDiGraph(nx.subgraph_view(g, filter_edge=lambda u, v, k: u > 0))
print("Edges in filtered graph:", len(gg.edges))
```
`requirements.txt`
```
nuitka==1.5.5
networkx==2.8.8
```
```
>python.exe main.py
Edges in original graph: 2
Edges in filtered graph: 1
```

Same issue on Windows with **Commercial** license, no Segmentation fault output but program terminated silently.

```
(.venv) > python -m nuitka --version
1.5.5
Commercial: 2.0.0
Python: 3.10.4 (tags/v3.10.4:9d38120, Mar 23 2022, 23:13:41) [MSC v.1929 64 bit (AMD64)]
Flavor: CPython Official
Executable: C:\Users\anton\Documents\nuitka_nx\.venv\Scripts\python.exe
OS: Windows
Arch: x86_64
WindowsRelease: 10
Version C compiler: C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.34.31933\bin\HostX64\x64\cl.exe (cl 14.3).
```

```
(.venv) > python -m nuitka --standalone main.py
Nuitka-Options:INFO: Used command line options: --standalone main.py
Nuitka:INFO: Starting Python compilation with Nuitka '1.5.5' on Python '3.10' commercial grade '2.0.0'.
Nuitka:INFO: Completed Python level compilation and optimization.
Nuitka:INFO: Generating source code for C backend compiler.
Nuitka:INFO: Running data composer tool for optimal constant value handling.
Nuitka:INFO: Running C compilation via Scons.
Nuitka-Scons:INFO: Backend C compiler: cl (cl 14.3).
Nuitka-Scons:INFO: Backend linking program with 281 files (no progress information available).
Nuitka-Scons:INFO: Compiled 281 C files using clcache with 0 cache hits and 281 cache misses.
Nuitka:INFO: Keeping build directory 'main.build'.
Nuitka:INFO: Successfully created 'main.dist\main.exe'.

(.venv) > .\main.dist\main.exe
Edges in original graph: 2
(.venv) PS >
```

As workaround in this particular case we can use `g.copy()` instead of calling constructor

```
import networkx as nx

g = nx.MultiDiGraph()
g.add_edge(0, 1)
g.add_edge(1, 2)
print("Edges in original graph:", len(g.edges))

gg = nx.subgraph_view(g, filter_edge=lambda u, v, k: u > 0).copy()
print("Edges in filtered graph:", len(gg.edges))
```

but similar issues happens in different scenarios with `networkx`, so suppose it's better to investigate it as is.

**Comment by AntonOvsyannikov** [comment]
The seg fault can also be reproduced on this snippet
```python
import networkx as nx
g = nx.MultiDiGraph()
g.add_edge(0, 2)
g.add_edge(2, 1)
g = nx.transitive_reduction(g)
```

`Dockerfile`
```
FROM python:3.10
WORKDIR /srv
ENV PYTHONUNBUFFERED=1

RUN apt update
RUN apt install patchelf

RUN printf 'nuitka==1.5.5\nnetworkx==2.8.8' >> requirements.txt


RUN python -m pip install -r requirements.txt

RUN printf 'import networkx as nx\n\
g = nx.MultiDiGraph()\n\
g.add_edge(0, 2)\n\
g.add_edge(2, 1)\n\
g = nx.transitive_reduction(g)' >> main.py

RUN python -m nuitka --standalone main.py

CMD "./main.dist/main.bin"
```

**Comment by kayhayen** [comment]
Ok, this is a first. 

There is an infitite recursion going on with the `impl_networkx$utils$decorators$$$function__10___call__$$$function__2_func`  being on the stack. The segfault is eventually when trying to allocate an empty dictionary, but it seems it's maybe just running out of memory in `PyDict_New`.

**Comment by kayhayen** [comment]
Ok, so this is in a decorator, that I think I have had to modify in the past, so it works with Nuitka. But now it no longer does. Maybe it's something stupid, likely it works only once, and then it would not, and we didn't notice it, or this only worked on an older version, but I swear it was necessary and working at some point. But well, it's not. With `--full-compat` the recursion error is visible, but of course that is not recommended, since it adds a lot of stuff.

But I believe that should make it such, that the fix is probably relatively easy, onec I wrap my head around how to achieve what they are trying there.

**Comment by AntonOvsyannikov** [comment]
@kayhayen hello, sorry, are there any plans for this issue? Can we help with something here?

**Comment by kayhayen** [comment]
So, this is a bit of a rough one. There are libraries like `networkx`, but also others, which want and do play heavy tricks with the source code. I am refering to this part e.g.

```
    def compile(self, f):
        """Compile the decorated function.

        Called once for a given decorated function -- collects the code from all
        argmap decorators in the stack, and compiles the decorated function.

        Much of the work done here uses the `assemble` method to allow recursive
        treatment of multiple argmap decorators on a single decorated function.
        That flattens the argmap decorators, collects the source code to construct
        a single decorated function, then compiles/executes/returns that function.

        The source code for the decorated function is stored as an attribute
        `_code` on the function object itself.

        Note that Python's `compile` function requires a filename, but this
        code is constructed without a file, so a fictitious filename is used
        to describe where the function comes from. The name is something like:
        "argmap compilation 4".

        Parameters
        ----------
        f : callable
            The function to be decorated

        Returns
        -------
        func : callable
            The decorated file

        """
        sig, wrapped_name, functions, mapblock, finallys, mutable_args = self.assemble(
            f
        )

        call = f"{sig.call_sig.format(wrapped_name)}#"
        mut_args = f"{sig.args} = list({sig.args})" if mutable_args else ""
        body = argmap._indent(sig.def_sig, mut_args, mapblock, call, finallys)
        code = "\n".join(body)

        locl = {}
        globl = dict(functions.values())
        filename = f"{self.__class__} compilation {self._count()}"
        compiled = compile(code, filename, "exec")
        exec(compiled, globl, locl)
        func = locl[sig.name]
        func._code = code
        return func
```

Some of these are useless, but far from all of them are. So I was using bandaids, and probably will continue here to use one, but I am very worried, that the real solution is to compile time capture these uses of `exec` and then at run time to have the exec result, including compiled functions just ready. That's a very painful change. I will have to do something like that one day, but lets face it, that will take a while.

Another approach would be to remove useless decorations in networkx, e.g. by doing this:

```
# @not_implemented_for("undirected")
def transitive_reduction(G):
```

Assuming the check is not really needed, we could just not do it and low and behold, the decorator issue goes away for our example. Adapting them one by one is clearly an option. My previous change, I am pretty sure, makes some of them work fine already. 

So what I will try and do now, is to make the argmap usage of `not_implemented_for` use `functools.wraps` and mark these functions for inlining in the future. For Nuitka, the real optimization that we should be aiming for is to inline the decorator, and have the decorator then inline its function argument. In that way, Nuitka would be the perfect solution. 

As another topic, for people who don't do weird things to avoid it, the time being, `functools.wraps` is something that could get an optimized replacement in Nuitka too. I have noticed, in other contexts, that it presents a huge slowdown for calls. It's call mechanics cannot optimize on argument numbers, etc. and do silly merges and slow calls, which is what drove networkx to do this in the first place.

Unfortunately, that means, I will probably have a solution for your example real soon now. But I guess, we should use the factory/staging approach to find out, if you have similar issues in the rest of your software, and it might take a few iterations, but I am definitely looking to do that with you.


This appears to be a list of the decorators:

```
    See Also
    --------
    not_implemented_for
    open_file
    nodes_or_number
    random_state
    py_random_state
    networkx.community.quality.require_partition
    require_partition
```

I believe some look as if they are more for testing purposes or little used.

The real solution, namely cross module forced inling of function calls, is something that will take a bit longer, and I don't think you should have to wait for it and I feel, the `exec` from above may never quite be doable without help. Capturing the `exec` calls, and making that use compiled code, is also an option, but I will resort to it only if the inlining fails for too long. 

The workarounds, I expect should be very doable in the near future. I just honestly have to say, that their specific decoration tool is a bit of a mystery to me still. But I guess, I can extract it from examples.

That said, having "checking" decorators only check in Python but not in deployment, is probably something in the veins of `anti-bloat` and we should eventually figure out package specific options, where you can e.g. opt-in to have `@not_implemented_for` just do nothing, but for now I would find it to gross.

**Comment by mishamsk** [comment]
@kayhayen hey! thanks for looking into that. I am working with @AntonOvsyannikov on the same project, so I’ll chime in.

First of all, I think the `argmap` in general is a bit of an over engineered magic, although I myself prone to such stunts, I’d hope popular libraries would stray away from such tricks;-)

But even if a bit of wizardry is ok, using a monster like argmap on such a simple decorator as `not_implemented_for` is definitely excessive. I’d say that commenting it out in anti-bloat style is a perfect solution and would cover our needs for now for sure. We’d easily do a check prior to the call.

I did check the rest of the networkx API that we use or plan on using soon (specifically the `topological sort`), it seems like we are good, argmap is not used from what I can tell. Also now we’ll know which methods to avoid or reimplement ourselves.  

So in short - happy to try out a factory route with an actual inlined/altered implementation for some of the decorators, but would also appreciate an option to just comment the decorator out and forget about it. Maybe doing both in different branches. Whatever works for you.

If there is a way to “plugin in” anti-bloat like logic in our specific environment - would be awesome. We actually have another place to use such plugin - we import sqlfluff, which brings in pytest and simple anti-bloat breaks the lib. But if we could instruct nuitka to comment two lines in their codebase prior to compiling, like in some of your more advanced anti-bloat configs - that would work.

**Comment by kayhayen** [comment]
Thanks, I have updated both factory and staging (factory for commercial users) branches. Give it a shot. I expect it to be very stable at this time, the only issue could be hidden depdnencies in stdlib since I made an effort to expand non-automatic stdlib modules.

**Comment by kayhayen** [comment]
I had a bit of a struggle there with Yaml and our autoformat breaking the code. I think it always worked on factory, but repeated formatting of the file was not stable, apparently. Had to fix that. I am pushing out regular 1.5.7 with this and a bunch of other changes. Actually one of the larger hotfixes in a while, as I delayed it for this fix which seems good even if you later need more.

**Comment by kayhayen** [comment]
I am closing this one, because 1.5.7 doesn't reproduce it anymore. It will hit commercial tomorrow, likely. Feel free to come back here, if you have additional findings.

**Comment by mishamsk** [comment]
Hi, sure, no prob. We are currently busy prepping a new release which has the related features temporarily disabled, but I’ll get back to re-enabling them next week if not earlier. I’ll report if we encounter any issues
